### ISSUE TITLE
Http Cache in v3 caches mutations whereas v2 does not - is that a feature?

### ISSUE THREAD

**Issue by soldierinwhite** [apollographql/apollo-kotlin#5012]
### Question

According to the [v2 docs for HTTP Cache](https://www.apollographql.com/docs/kotlin/v2/essentials/http-cache):
`IMPORTANT: Caching is provided only for query operations. It isn't available for mutation operations.`

However, in v3, nothing of the sort is mentioned, but neither is it mentioned in v2 to v3 migration docs. When testing mutations, they now return cached responses in v3 according to the global cache policy. That means client mutations will break as soon as you migrate and you haven't explicitly changed, and there is no indication of that happening in the docs.

Further, caching a mutation does not really make sense to me. How can I intend to mutate something in the server, but then I am satisfied with not doing that and behaving as though it did it with the answer stored in cache from the previous time I did the mutation. Is the intention that clients should be adding `httpDoNotStore()` manually on every mutation to avoid this even though that seems like you would always want it to be the case?

So 1, if we do not add any changes, can we make this clear in the docs that the behaviour for mutations has changed
And 2, does it make sense to cache mutations at all in v3, and could we at least not make it the default? The use case for when caching mutations is helpful seems like an edge case at best.

**Comment by BoD** [comment]
Hi! Thanks for reporting this.

I agree that it doesn't make much sense to cache mutations - or at least there are no use case that I can think of where you would want to do that.

The default policy [is actually](https://github.com/apollographql/apollo-kotlin/blob/aa847f427145906f5a1a215281db22dcabe973e1/libraries/apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/HttpCacheExtensions.kt#L145) `CacheFirst` for queries and `NetworkOnly` (i.e. no cache) for mutations. But if you do set another policy (e.g. `NetworkFirst`) globally on your client, or on a specific call then it will be used as-is. Is it what you are doing?

**Comment by soldierinwhite** [comment]
> Hi! Thanks for reporting this.
> 
> I agree that it doesn't make much sense to cache mutations - or at least there are no use case that I can think of where you would want to do that.
> 
> The default policy [is actually](https://github.com/apollographql/apollo-kotlin/blob/aa847f427145906f5a1a215281db22dcabe973e1/libraries/apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/HttpCacheExtensions.kt#L145) `CacheFirst` for queries and `NetworkOnly` (i.e. no cache) for mutations. But if you do set another policy (e.g. `NetworkFirst`) globally on your client, or on a specific call then it will be used as-is. Is it what you are doing?

Yes, we are using NetwokrFirst since we just implemented HttpCache and wanted as little change to current behaviour as possible, just covering network loss with caching. But I did not expect NetworkFirst to be set for mutations as well. Is it possible to set different policies for queries and mutations with the client builder?

**Comment by BoD** [comment]
Thank you for confirming.

I think it would probably be reasonable to change the behavior so that mutations are never cached (and clearly document it).

In the meantime, you can use an interceptor to automatically set a `NetworkOnly` policy for mutations. Something like:

```kotlin
object HttpCachePolicyInterceptor : ApolloInterceptor {
    override fun <D : Operation.Data> intercept(
        request: ApolloRequest<D>,
        chain: ApolloInterceptorChain
    ): Flow<ApolloResponse<D>> {
        return chain.proceed(
            if (request.operation is Mutation<*>) {
                request.newBuilder().httpFetchPolicy(HttpFetchPolicy.NetworkOnly).build()
            } else {
                request
            }
        )
    }
}
```

Then set the interceptor on your client (make sure to set it before the http cache):

```kotlin
val apolloClient = ApolloClient.Builder()
        .serverUrl(...)
        .addInterceptor(HttpCachePolicyInterceptor)
        .httpCache(...)
        .build()

```

**Comment by BoD** [comment]
[The change](https://github.com/apollographql/apollo-kotlin/pull/5014) will be available in the 4.0 release.
